src/platform/Darwin/DnssdContexts.cpp

```diff
@@ -0,0 +1,426 @@
+        ips.erase(ips.begin());
+        callback(context, &service.second, Span<Inet::IPAddress>(ips.data(), ips.size()), CHIP_NO_ERROR);
+        MdnsContexts::GetInstance().Remove(this);
+        return;
+    }
+}
+
+CHIP_ERROR ResolveContext::OnNewAddress(uint32_t interfaceId, const struct sockaddr * address)
+{
+    chip::Inet::IPAddress ip;
+    ReturnErrorOnFailure(chip::Inet::IPAddress::GetIPAddressFromSockAddr(*address, ip));
+    addresses[interfaceId].push_back(ip);
+
+#ifdef CHIP_DETAIL_LOGGING
+    char addrStr[INET6_ADDRSTRLEN];
+    ip.ToString(addrStr, sizeof(addrStr));
+    ChipLogDetail(DeviceLayer, "Mdns: %s interface: %" PRIu32 " ip:%s", __func__, interfaceId, addrStr);
+#endif // CHIP_DETAIL_LOGGING
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ResolveContext::OnNewLocalOnlyAddress()
+{
+    sockaddr_in6 sockaddr;
+    memset(&sockaddr, 0, sizeof(sockaddr));
+    sockaddr.sin6_len    = sizeof(sockaddr);
+    sockaddr.sin6_family = AF_INET6;
+    sockaddr.sin6_addr   = in6addr_loopback;
+    sockaddr.sin6_port   = htons((unsigned short) services[kDNSServiceInterfaceIndexLocalOnly].mPort);
+
+    return OnNewAddress(kDNSServiceInterfaceIndexLocalOnly, reinterpret_cast<struct sockaddr *>(&sockaddr));
+}
+
+bool ResolveContext::HasAddress()
+{
+    for (auto & interfaceAddresses : addresses)
+    {
+        if (interfaceAddresses.second.size())
+        {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void ResolveContext::OnNewInterface(uint32_t interfaceId, const char * fullname, const char * hostnameWithDomain, uint16_t port,
+                                    uint16_t txtLen, const unsigned char * txtRecord)
+{
+    ChipLogDetail(DeviceLayer, "Mdns : %s hostname:%s fullname:%s interface: %" PRIu32, __func__, hostnameWithDomain, fullname,
+                  interfaceId);
+
+    DnssdService service = {};
+    service.mPort        = ntohs(port);
+    service.mInterface   = Inet::InterfaceId(interfaceId);
+
+    auto hostname = GetHostNameWithoutDomain(hostnameWithDomain);
+    Platform::CopyString(service.mHostName, hostname.c_str());
+    Platform::CopyString(service.mName, fullname);
+
+    GetTextEntries(service, txtRecord, txtLen);
+
+    addresses.insert(std::pair<uint32_t, std::vector<Inet::IPAddress>>(interfaceId, std::vector<Inet::IPAddress>()));
+    services.insert(std::pair<uint32_t, DnssdService>(interfaceId, service));
+}
+
+bool ResolveContext::HasInterface()
+{
+    return services.size();
+}
+
+void ResolveContext::RemoveInterfaces()
+{
+    for (auto & service : services)
+    {
+        size_t count = service.second.mTextEntrySize;
+        for (size_t i = 0; i < count; i++)
+        {
+            const auto & textEntry = service.second.mTextEntries[i];
+            free(const_cast<char *>(textEntry.mKey));
+            free(const_cast<uint8_t *>(textEntry.mData));
+        }
+        Platform::MemoryFree(const_cast<TextEntry *>(service.second.mTextEntries));
```

Comment:
The allocator mixing here is a bit weird, but it matches the allocation sites, so ok....